### DISCUSSION TITLE
Too many clang warnings

### DISCUSSION THREAD

**Discussion by gvanem** [curl/curl#9319] [General]
When compiling any project that uses libcurl, there are tons of warning from clang/clang-cl. Like in PyCurl:
![clang-warnings](https://user-images.githubusercontent.com/945271/184633749-ddba1f36-e4df-48d3-9930-38fd714b4169.png)

Since the `<curl/header.h> ` part `/**` ... is interpreted as a Doxygen tag and `@haxx` is not a valid Doxygen command.
Compiling with `-Wno-documentation-unknown-command` is a fix. 

But sometimes when one need to validate Doxygen tags, this warning is useful. 
So could the headers start with `/*---` instead perhaps? 

```
c:\>clang -v
clang version 14.0.6
Target: x86_64-pc-windows-msvc
```

**Comment by bagder** [comment]
Yikes, what makes clang interpret the comments? And what makes it dictate the rules for comment contents?

I count 978 instances of `/**` in curl.

**Reply by bagder** [reply]
But I guess only public headers matter, so 13 instances.

**Comment by gvanem** [comment]
Clang is pretty zealous when it comes to warnings. The docs on this one:
https://releases.llvm.org/14.0.0/tools/clang/docs/DiagnosticsReference.html#wdocumentation-unknown-command

**Reply by bagder** [reply]
Are these warning then enabled by default or what makes it decide that it should check for doxygen commands in the first place?

**Reply by gvanem** [reply]
I only get this warning on `-Wall`.

**Reply by emanuele6** [reply]
@gvanem
I don't get this warning with `-Wall`, not even with `-Wextra`.
I need to use `-Weverything` to get it. (`make clean; make CC=clang CFLAGS=-Weverything`)

---
```
bash-5.1$ clang --version
clang version 14.0.6
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

**Reply by gvanem** [reply]
@emanuele6 I'm on Windows using `clang-cl`. Then there is a difference; according to [this](https://clang.llvm.org/docs/UsersManual.html?highlight=weverything#id10), a `-Wall == -Weverything`.

**Reply by emanuele6** [reply]
I see! So `/Wall` on the windows compiler is equivalent to `-Weverything`, `/W4` is equivalent to `-Wall -Wextra`, and `/W1`, `/W2`, `/W3` are all equivalent to `-Wall` =)

**Reply by bagder** [reply]
Hm?

[Don’t put -Weverything in your build flags](https://quuxplusone.github.io/blog/2018/12/06/dont-use-weverything/)

**Reply by bagder** [reply]
With clang 14.0.6-2 on Debian Linux, I can't even see it with `-Weverything` on a small sample program including the curl headers...
~~~c
#include <curl/curl.h>
#include <curl/header.h>

int main(void)
{
  CURL *easy = curl_easy_init();
  curl_easy_setopt(easy, CURLOPT_URL, "localhost");
  return 0;
}

**Reply by emanuele6** [reply]
I am able to get it if I include `curl.h` using `#include "path"`:
```
bash-5.1$ ls -a
.  ..  testfile.c
bash-5.1$ cat testfile.c
#include "/usr/include/curl/curl.h"

int main(void)
{
  CURL *easy = curl_easy_init();
  curl_easy_setopt(easy, CURLOPT_URL, "localhost");
  return 0;
}
bash-5.1$ make CC=clang CFLAGS='-Weverything -lcurl' testfile
clang -Weverything -lcurl    testfile.c   -o testfile
In file included from testfile.c:1:
In file included from /usr/include/curl/curl.h:3097:
/usr/include/curl/header.h:10:55: warning: unknown command tag name [-Wdocumentation-unknown-command]
 * Copyright (C) 2018 - 2022, Daniel Stenberg, <daniel@haxx.se>, et al.
                                                      ^~~~~
testfile.c:6:3: warning: disabled expansion of recursive macro [-Wdisabled-macro-expansion]
  curl_easy_setopt(easy, CURLOPT_URL, "localhost");
  ^
/usr/include/curl/curl.h:3109:44: note: expanded from macro 'curl_easy_setopt'
#define curl_easy_setopt(handle,opt,param) curl_easy_setopt(handle,opt,param)
                                           ^
2 warnings generated.
```
I think `clang` (and probably also other compilers) only check warnings on header files included with `"path"` to avoid reporting problems on third party code.

**Reply by gvanem** [reply]
A `"/usr/include"` or not is related to warnings on [system-headers](https://clang.llvm.org/docs/UsersManual.html?highlight=weverything#controlling-diagnostics-in-system-headers).

**Reply by bagder** [reply]
ah of course

**Comment by dfandrich** [comment]
Does it interpret only "/** " (with a following space) specially or "/*****…"
(any number of *) ?  If the former, there aren't many cases of this. The
Doxygen documentation says that the "/***********" style isn't
interpreted specially by default, so hopefully, there's a clang option to make
it work like Doxygen.
